src/js/structs/CompositeState.js

```diff
@@ -1,53 +1,27 @@
 import NodesList from "./NodesList";
 import Node from "./Node";
+import Util from "../utils/Util";
 import ServicesList from "../../../plugins/services/src/js/structs/ServicesList";
 
-const BLANK_NODE = {
+const PRESERVE_KEYS = ["master_info"];
+const MISSING_HEALTH_NODE = {
   health: 3
 };
 
-const mergeData = function(newData, data) {
-  Object.keys(newData).forEach(function(key) {
-    if (Array.isArray(newData[key])) {
-      data[key] = mergeMesosArrays(newData, data, key);
-    } else if (typeof newData[key] === "object" && data[key]) {
-      // We need to recurse over any nested objects.
-      data[key] = mergeData(newData[key], data[key]);
-    } else {
-      // Any other type of value can be replaced.
-      data[key] = newData[key];
-    }
-  });
-
-  return data;
-};
+// replaces node data with previous node data plus health data
+const enrichNodeDataWithHealthData = (nodes, healthData) => {
+  return nodes.map(function(node) {
+    const matchedHealthNode = healthData[node.hostname] || MISSING_HEALTH_NODE;
 
-const mergeMesosArrays = function(newData, data, key) {
-  if (key === "frameworks" || key === "slaves") {
-    // We need to merge the objects within the frameworks and slaves arrays.
-    return mergeObjectsById(newData[key], data[key]);
-  } else {
-    // We can replace any other array.
-    return newData[key];
-  }
-};
-
-const mergeObjectsById = function(newData, data = []) {
-  // Merge the incoming data with the old data.
-  return newData.map(function(newDatum) {
-    const oldDatum = data.find(function(datum) {
-      return datum.id === newDatum.id;
-    });
-
-    // These objects don't need to be deeply merged.
-    return { ...oldDatum, ...newDatum };
+    return { ...node, ...matchedHealthNode };
```

Comment:
Let's please do this!

Reply:
Wow! This really helped much more than I expected. 60ms -> 2ms with fixtures

Reply:
WOW! 🎉  What if we do a mutation instead?

Reply:
I had done both changes at the same time. If create a new object but assign health field instead of spreading matchedHealthNode, it only shaves off a few ms. The new object and spreading `node` is the expensive part.

I changed the nodes and node health fixtures to 50,000 and it was still usable and these two methods were timing at 10ms or less

Reply:
Bam 💥